npeps/npep-187-ports-and-protocols.md

```diff
@@ -0,0 +1,245 @@
+    # protocol: "" # must be empty string: defaulted / omitempty
+```
+
+with the validation like so:
+- empty `protocols` list is not allowed
+- at least 1 field for each `protocol` element must be set
+- if `namedPort` is set, `protocol` must be unset
+- if `protocol` is TCP/UDP/SCTP, `port` must be set
+
+### Option2: Join by protocol
+
+```yaml
+protocols:
+  tcp:
+    - port:
+        number: 8080
+    - port:
+        range: 
+          start: 8080
+          end: 9090
+  udp:
+    - port:
+        number: 8080
+    - port:
+        number: 9090
+  namedPorts:
+    - http
+    - monitoring
+  icmp:
+    - type: 7
+      code: 3
+```
+
+In this case it makes more sense to also flatten the `port` match, because for TCP/UDP/SCTP protocols, port is the only option and is always required:
+
+```yaml
+protocols:
+  tcp:
+    - port: 8080
+    - portRange: 
+        start: 8080
+        end: 9090
+  udp:
+    - port: 8080
+    - port: 9090
+  namedPorts:
+    - http
+    - monitoring
+  icmp:
+    - type: 7
+      code: 3
+```
+
+with the validation like so:
+- empty `protocols` list is not allowed
+- at least 1 field of `protocol` must be set
+- each protocol list must not be empty
+
+This option is easier to write, takes less space and simplifies the validation, but doesn't allow defaulting protocol to TCP 
+(which is the most popular case). On the other hand, I am not convinced that defaulting protocol to TCP is
+a) a good idea (less transparency on which protocol is used)
+b) much simpler than adding `tcp:`
+
+To compare, the first option for `tcp:8080` is
+```yaml
+protocols:
+  - port: 
+      number: 8080
+```
+and the second is
+```yaml
+protocols:
+  tcp:
+    - port: 8080
+```
+
+And if you want to match 2 ports:
+
+```yaml
+protocols:
+  - port: 
+      number: 8080
+  - port:
+      number: 9090
+```
+vs
+```yaml
+protocols:
+  tcp:
+    - port: 8080
+    - port: 9090
+```
+
+Now add UDP 53:
+
+```yaml
+protocols:
+  - port: 
+      number: 8080
+  - protocol: UDP
+    port:
+      number: 53
+```
+vs
+```yaml
+protocols:
+  tcp:
+    - port: 8080
+  udp:
+    - port: 53
+```
+
+### New protocols
+
+Using ICMP as a potential example for a pure protocol match (`protocol: ICMP`) or protocol with extra fields (e.g. ICMP type and code),
+a future extension could look like this:
+
+Option 1:
+```yaml
+protocols:
+  # option 1:  A little inconsistent with port match as it has an extra `icmp` field instead of flat parameters
+  - protocol: ICMP
+    icmp:
+      type: 7
+      code: 3
+  # option 2: Flat parameters, may conflict with the future protocol that also have type or code
+  - protocol: ICMP
+    type: 7
+    code: 3
+  # option 3: Same as option 1, but without a discriminator (protocol)
+  - icmp:
+      type: 7
+      code: 3
+```
+
+Option 2:
+```yaml
+protocols:
+  icmp:
+    - type: 7
+      code: 3
+```
+
+A likely option is that we will start with the `ICMP: match all` option, as more details matching is not very popular,
```

Comment:
i think a lot of people want to block pings but not `host unreachable`

Reply:
I think v6 would be `icmp6`. I'm looking at the codes and type and they don't line up:

* https://en.wikipedia.org/wiki/Internet_Control_Message_Protocol
* https://en.wikipedia.org/wiki/ICMPv6

Reply:
> i think a lot of people want to block pings but not host unreachable

Not that we need to figure it out now, but ICMP host unreachable is typically a response packet to an outgoing UDP/TCP packet.  It's best handled as part of the UDP/TCP session from a firewall point of view (I think this is what Linux conntrack does, for example).